x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -528,24 +539,59 @@ private void findActiveApiKeys(final BoolQueryBuilder boolQuery, ActionListener<
             .request();
         securityIndex.checkIndexVersionThenExecute(listener::onFailure,
             () -> ScrollHelper.fetchAllByEntity(client, request, listener,
-                (SearchHit hit) -> hit.getId()));
+                        (SearchHit hit) -> {
+                            Map<String, Object> source = hit.getSourceAsMap();
+                            String name = (String) source.get("name");
+                            String id = hit.getId();
+                            Long creation = (Long) source.get("creation_time");
+                            Long expiration = (Long) source.get("expiration_time");
+                            Boolean invalidated = (Boolean) source.get("api_key_invalidated");
+                            String username = (String) ((Map<String, Object>) source.get("creator")).get("principal");
+                            String realm = (String) ((Map<String, Object>) source.get("creator")).get("realm");
+                            return new ApiKey(name, id, Instant.ofEpochMilli(creation),
+                                    (expiration != null) ? Instant.ofEpochMilli(expiration) : null, invalidated, username, realm);
+                        }));
     }
 
     private void findActiveApiKeyForApiKeyName(String apiKeyName, ActionListener<Collection<String>> listener) {
+        findApiKeyForApiKeyName(apiKeyName, true, ActionListener.wrap(apiKeyInfos -> {
+            listener.onResponse(apiKeyInfos.stream().map(o -> o.getId()).collect(Collectors.toList()));
+        }, listener::onFailure));
+    }
+
+    private void findApiKeyForApiKeyName(String apiKeyName, boolean onlyActiveApiKeys, ActionListener<Collection<ApiKey>> listener) {
         final SecurityIndexManager frozenSecurityIndex = securityIndex.freeze();
         if (frozenSecurityIndex.indexExists() == false) {
             listener.onResponse(Collections.emptyList());
         } else if (frozenSecurityIndex.isAvailable() == false) {
             listener.onFailure(frozenSecurityIndex.getUnavailableReason());
         } else {
             final BoolQueryBuilder boolQuery = QueryBuilders.boolQuery()
-                .filter(QueryBuilders.termQuery("doc_type", "api_key"))
-                .filter(QueryBuilders.termQuery("api_key_invalidated", false));
+                .filter(QueryBuilders.termQuery("doc_type", "api_key"));
+            if (onlyActiveApiKeys) {
+                boolQuery.filter(QueryBuilders.termQuery("api_key_invalidated", false));
+            }
             if (Strings.hasText(apiKeyName)) {
                 boolQuery.filter(QueryBuilders.termQuery("name", apiKeyName));
             }
 
-            findActiveApiKeys(boolQuery, listener);
+            findApiKeys(boolQuery, listener);
+        }
+    }
+
+    private void findApiKeysForApiKeyId(String apiKeyId,
```

Comment:
We should probably have the same behavior here and in `findApiKeyForApiKeyName`. Either both need to return the `invalidated` (and `expired` - see comments above ) tokens, or neither should. I see that by default this is what we do as in `getApiKeyForApiKeyName` we call the `findApiKeyForApiKeyName` with `onlyActiveApiKeys` set to `false` , but do we have a use case for "getting active api keys by name" and not for "getting active api keys by id" ?

Reply:
Thanks for the explanation :) 

> during the creation of API keys, for checking if the API key name is already used by other API keys, the check includes expired and invalidated keys

I think we should change this. I can see a valid use case for invalidating an api key by name so that I can create a new one with that name and I shouldn't have to wait for up to a week until the invalidated one gets deleted.